crates/red_knot_python_semantic/src/types/signatures.rs

```diff
@@ -0,0 +1,480 @@
+#![allow(dead_code)]
+use super::{definition_expression_ty, Type};
+use crate::semantic_index::definition::Definition;
+use crate::Db;
+use ruff_python_ast::{self as ast, name::Name};
+
+/// A typed callable signature.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) struct Signature<'db> {
+    parameters: Parameters<'db>,
+
+    /// Annotated return type (Unknown if no annotation.)
+    pub(crate) return_ty: Type<'db>,
+}
+
+impl<'db> Signature<'db> {
+    /// Return a todo signature: (*args: Todo, **kwargs: Todo) -> Todo
+    pub(crate) fn todo() -> Self {
+        Self {
+            parameters: Parameters::todo(),
+            return_ty: Type::Todo,
+        }
+    }
+
+    /// Return a typed signature from a function definition.
+    pub(super) fn from_function(
+        db: &'db dyn Db,
+        definition: Definition<'db>,
+        function_node: &'db ast::StmtFunctionDef,
+    ) -> Self {
+        let return_ty = function_node
+            .returns
+            .as_ref()
+            .map(|returns| {
+                if function_node.is_async {
+                    // TODO: generic `types.CoroutineType`!
+                    Type::Todo
+                } else {
+                    definition_expression_ty(db, definition, returns.as_ref())
+                }
+            })
+            .unwrap_or(Type::Unknown);
+
+        Self {
+            parameters: Parameters::from_parameters(
+                db,
+                definition,
+                function_node.parameters.as_ref(),
+            ),
+            return_ty,
+        }
+    }
+}
+
+/// The parameters portion of a typed signature.
+///
+/// The ordering of parameters is always as given in this struct: first positional-only parameters,
+/// then positional-or-keyword, then optionally the variadic parameter, then keyword-only
+/// parameters, and last, optionally the variadic keywords parameter.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(super) struct Parameters<'db> {
+    /// Parameters which may only be filled by positional arguments.
+    positional_only: Box<[ParameterWithDefault<'db>]>,
+
+    /// Parameters which may be filled by positional or keyword arguments.
+    positional_or_keyword: Box<[ParameterWithDefault<'db>]>,
+
+    /// The `*args` variadic parameter, if any.
+    variadic: Option<Parameter<'db>>,
+
+    /// Parameters which may only be filled by keyword arguments.
+    keyword_only: Box<[ParameterWithDefault<'db>]>,
+
+    /// The `**kwargs` variadic keywords parameter, if any.
+    keywords: Option<Parameter<'db>>,
```

Comment:
Have you considered an alternative representation where we store all `Parameters` in a single slice and instead use an enum for the parameter types. This has the downside that it allows to e.g. contain two `keywords` but makes iterating easier and reduces the overall allocations by a fair amount.

Reply:
It's not just about reducing allocations. I also just find using the parameter struct in rust awkward because I always have to remember the iteration order, chain over multiple slices (good bye `len`) for little good. I'm not very worried that we create too many invalid representations because we should construct this struct in exactly one location but we'll read it in many places. That's why I favor whatever design is more ergonomic when *using* it.

Reply:
The reason why we are considering changing this in the parser is to allow invalid parameter ordering to be represented for improved error recovery. For example, there's currently a  parser bug that it drops multiple kwargs argument because we can't represent it in the AST https://github.com/astral-sh/ruff/issues/14315

Reply:
> (good bye `len`)

I understand your point, but FWIW: we expose a `len()` method for `ast::Parameters`! It wouldn't be that hard to do similarly here.

Reply:
I agree with Alex that this representation serves a very different use case from the parser representation. While the parser representation should/must represent invalid signatures, for error recovery, `Signature` needs to serve _semantic_ use cases, like call checking and subtyping -- there is no sensible or meaningful interpretation of those semantic operations over an invalid signature. So when we encounter an invalid signature, we need to emit a diagnostic and use a fallback signature, we cannot allow this struct to represent invalid signatures.

So I don't think that either error recovery or the parser's likely future change is an argument for changing to a flat representation here.

That leaves it as a question of efficiency, ergonomics, and robustness.

Efficiency argues in favor of a flat representation, for less allocation, though it's also true that we could reduce the allocations a lot with a smallvec-like approach.

Robustness argues in favor of a structured representation, so that we can't represent invalid signatures. Though I also agree that we can minimize the impact here because these should be immutable and only constructed in one place.

Ergonomics is tricky. For some use cases a flat representation has better ergonomics, but I'm not clear yet if that will be true for the more complex semantic use cases we will need here -- there is an ergonomic cost to requiring client code to implicitly "consider" the possibility of invalid signatures that we should never represent. Though I think ultimately the ergonomic issues in either direction can be handled with API -- we can provide API for flattened iteration of parameters, and length, over a structured internal representation, and we can expose structured API that internally asserts validity, over a flat internal representation.

So all in all, the tradeoffs here are not clear to me :)

Reply:
I also think the ergonomics issue can be handled with API. We already have a [rich API](https://github.com/astral-sh/ruff/blob/5fcf0afff4a0242a72759a3c3cfaa314b53fe8f3/crates/ruff_python_ast/src/nodes.rs#L3507-L3568) for `ast::Parameters` that allows you to directly iterate over the parameters, query the number of parameters, and more. We can easily do the same for this struct.

And as Carl says, for many use cases in a type checker such as questions of subtyping, the structured representation should really be more ergonomic than the proposed flat representation.

Reply:
I don't mind deferring the decision if representing erroneous parameters is intentionally not allowed. 

> I also think the ergonomics issue can be handled with API. We already have a [rich API](https://github.com/astral-sh/ruff/blob/5fcf0afff4a0242a72759a3c3cfaa314b53fe8f3/crates/ruff_python_ast/src/nodes.rs#L3507-L3568) for ast::Parameters that allows you to directly iterate over the parameters, query the number of parameters, and more.

Yeah, we can do that but it leads to a lot of code and less cache locality. I'm a fan of representing invalid state but it always comes at a cost and it's unclear to me if it's justified here, considering it isn't something we construct often. Anyway, let's defer this to later when we have more usages that can inform our decision making.

Reply:
Thanks for the good discussion here. Will leave it as is for now, but keep this in mind as a possible point to revisit in future.

(One thing that may be relevant at least to the performance considerations is that I suspect in practice _most_ functions have only positional-or-keyword arguments, which means even with the structured representation there would be only one allocation and one contiguous array of parameters.)